cpp/src/hash/hashing.cu

```diff
@@ -301,18 +473,53 @@ hash_partition_table(table_view const& input,
                            num_partitions * sizeof(size_type),
                            cudaMemcpyDeviceToHost,
                            stream));
+                           
+  if (num_partitions <= 1024) {
```

Comment:
why 1024?

Reply:
Sorry, a bit of context is needed here. I found that the optimizations in this PR scale poorly and eventually run out of shared memory as the number of partitions is increased. In this graph, you can see that the speedup from this optimization (yellow line) actually goes negative past 600 partitions.

The 1024 here is to fallback to the unoptimized base implementation when the number of partitions exceeds this. For this PR, 600 would be a better choice, but I had set it to 1024 for the follow-up PR https://github.com/rapidsai/cudf/pull/3808

I'll add a comment explaining the fallback and the choice of the number.

![image](https://user-images.githubusercontent.com/5580056/73039710-00aba500-3e0c-11ea-863a-c357eb5bf0a8.png)